lib/eknvfs/ekn-vfs.c

```diff
@@ -0,0 +1,340 @@
+
+#define EKN_URI "ekn"
+#define EKN_SCHEME_LEN 6
+
+struct _EknVfs
+{
+  GVfs parent;
+};
+
+typedef struct
+{
+  GSList *shards;         /* EosShardShardFile list */
+
+  GHashTable *extensions; /* (uri, GVfs *) table */
+  gchar     **schemes;    /* Schemes suported by all GVfs in extensions table */
+  GVfs       *local;      /* see g_vfs_get_local () */
+} EknVfsPrivate;
+
+enum
+{
+  PROP_0,
+
+  PROP_SHARDS,
+  N_PROPERTIES
+};
+
+static GParamSpec *properties[N_PROPERTIES];
+
+G_DEFINE_DYNAMIC_TYPE_EXTENDED (EknVfs,
+				ekn_vfs,
+				G_TYPE_VFS, 0,
+				G_ADD_PRIVATE_DYNAMIC (EknVfs))
+
+#define EKN_VFS_PRIVATE(d) ((EknVfsPrivate *) ekn_vfs_get_instance_private((EknVfs*)d))
+
+static void
+ekn_vfs_extension_points_init (EknVfs *self)
+{
+  EknVfsPrivate *priv = EKN_VFS_PRIVATE (self);
+  GIOExtensionPoint *epoint;
+  GList *l;
+
+  /* Hash table to quickly get the extension point with higher priority that
+   * supports a uri scheme.
+   */
+  priv->extensions = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                            NULL, g_object_unref);
+
+  /* Get GVfs extension point */
+  if (!(epoint = g_io_extension_point_lookup (G_VFS_EXTENSION_POINT_NAME)))
+    return;
+
+  /* Iterate over the list of Gvfs extension points */
+  for (l = g_io_extension_point_get_extensions (epoint); l; l = g_list_next (l))
+    {
+      GType type = g_io_extension_get_type (l->data);
+      gchar **schemes;
+      GVfs *vfs;
+      gint i;
+
+      /* Skip ourself and uninstantiable extensions points */
+      if (type == EKN_TYPE_VFS || !(vfs = g_object_new (type, NULL)))
+        continue;
+
+      /* Get all the native schemes suported by the extension point */
+      schemes = (gchar **) (* G_VFS_GET_CLASS (vfs)->get_supported_uri_schemes) (vfs);
+
+      /* Add them in out hash table */
+      for (i = 0; schemes[i]; i++)
+        {
+          if (!g_hash_table_lookup (priv->extensions, schemes[i]))
+            g_hash_table_insert (priv->extensions, schemes[i], g_object_ref (vfs));
+        }
+
+      g_object_unref (vfs);
+    }
+}
+
+static void
+ekn_vfs_init (EknVfs *self)
+{
+  EknVfsPrivate *priv = EKN_VFS_PRIVATE (self);
+  gchar **schemes;
+  guint length;
+
+  /* Init priv->extensions hash table */
+  ekn_vfs_extension_points_init (self);
+
+  /* Get suported all uri schemes */
+  schemes = (gchar **) g_hash_table_get_keys_as_array (priv->extensions, &length);
+  priv->schemes = g_realloc (schemes, (length + 2) * sizeof (gchar *));
+
+  /* And append our custom scheme at the end */
+  priv->schemes[length]   = EKN_URI;
+  priv->schemes[++length] = NULL;
+
+  /* Get local vfs */
+  priv->local = g_vfs_get_local ();
+}
+
+static void
+ekn_vfs_finalize (GObject *self)
+{
+  EknVfsPrivate *priv = EKN_VFS_PRIVATE (self);
+
+  g_free (priv->schemes);
+  g_hash_table_unref (priv->extensions);
+
+  if (priv->shards)
+    g_slist_free_full (priv->shards, g_object_unref);
```

Comment:
No need to check for `NULL`: a `NULL` list is still a valid list, so `g_slist_free_full()` is `NULL`-safe.